library/Faktory/Ent/Batch.hs

```diff
@@ -0,0 +1,129 @@
+--
+-- Batches allow multiple Jobs to be enqueued as a group, with a description
+-- (visible in the admin UI) and Jobs attached to run on completion of all Jobs
+-- within the group (always, or only if all were successful).
+--
+-- Usage:
+--
+-- @
+-- -- Build a Job to run at completion of the Batch. Arguments are the same as
+-- -- you would pass to 'perform' the Job.
+-- onComplete <- buildJob mempty producer myJob
+--
+-- 'runBatchT' ('complete' onComplete <> 'description' "My Batch") producer $ do
+--   -- Use 'batchPerform' instead of 'perform'
+--   void $ 'batchPerform' mempty producer myBatchedJob1
+--   void $ 'batchPerform' mempty producer myBatchedJob2
+-- @
+--
+-- __/NOTE__: This module does not support batched Jobs dynamically adding more
+-- Jobs to the Batch. PRs welcome.
+--
+module Faktory.Ent.Batch
+    (
+    -- * Options
+      BatchOptions
+    , description
+    , complete
+    , success
+
+    -- * Running
+    , runBatchT
+    , BatchT
+    , batchPerform
+    )
+where
+
+import Faktory.Prelude
+
+import Control.Monad.Reader
+import Data.Aeson
+import Data.Aeson.Casing
+import Data.ByteString.Lazy as BSL
+import Data.Semigroup (Last(..))
+import Data.Semigroup.Generic
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Faktory.Client
+import Faktory.Job
+import Faktory.Producer
+import GHC.Generics
+import GHC.Stack
+
+newtype BatchT a = BatchT (ReaderT BatchId IO a)
+  deriving newtype (Functor, Applicative, Monad, MonadIO, MonadReader BatchId)
+
+newtype BatchId = BatchId Text
+  deriving newtype ToJSON
+
+data BatchOptions arg = BatchOptions
+  { boDescription :: Maybe (Last Text)
+  , boSuccess :: Maybe (Last (Job arg))
+  , boComplete :: Maybe (Last (Job arg))
+  }
+  deriving stock Generic
+  deriving Semigroup via GenericSemigroupMonoid (BatchOptions arg)
+
+instance ToJSON arg => ToJSON (BatchOptions arg) where
+  toJSON = genericToJSON $ aesonPrefix snakeCase
+  toEncoding = genericToEncoding $ aesonPrefix snakeCase
+
+description :: Text -> BatchOptions arg
+description d = BatchOptions
+  { boDescription = Just $ Last d
+  , boSuccess = Nothing
+  , boComplete = Nothing
+  }
+
+complete :: Job arg -> BatchOptions arg
+complete job = BatchOptions
+  { boDescription = Nothing
+  , boSuccess = Nothing
+  , boComplete = Just $ Last job
+  }
+
+success :: Job arg -> BatchOptions arg
+success job = BatchOptions
+  { boDescription = Nothing
+  , boSuccess = Just $ Last job
+  , boComplete = Nothing
+  }
+
+runBatchT :: ToJSON arg => BatchOptions arg -> Producer -> BatchT a -> IO a
+runBatchT options producer (BatchT f) = do
+  bid <- newBatch producer options
+  result <- runReaderT f bid
+  result <$ commitBatch producer bid
+
+newtype CustomBatchId = CustomBatchId
+  { bid :: BatchId
+  }
+  deriving stock Generic
+  deriving anyclass ToJSON
+
+batchPerform
+  :: (HasCallStack, ToJSON arg) => JobOptions -> Producer -> arg -> BatchT JobId
+batchPerform options producer arg = do
+  bid <- ask
```

Comment:
Even though we're managing the batch, it might be worth having something that exposes the `BatchId` inside `runBatchT` to the client for logging purposes, e.g.:

```haskell
askBatchId :: BatchT BatchId
askBatchId = ask
```

And then defining `newtype (Eq, Show)` for `BatchId`

Reply:
It seems reasonable to add. I think I land on waiting to see if it's ever needed and adding it then, vs pre-exposing it in the module interface. I don't feel strongly -- do you want me to add it?

Reply:
I was thinking of where I would use it, e.g. managing progress for a long running job. I would probably have an independent resource that the completion handler updates, but I'd still probably wanna store off the batch id for debugging purposes so I know where to look in the admin UI if something gets stuck

Seems easy enough to add when we need it though

Reply:
We agree to just add it when/if we want it, but FWIW:

> managing progress ... independent resource

Faktory does have progress metrics itself, so hopefully we don't have to write that:

https://github.com/contribsys/faktory/wiki/Ent-Batches#status

> so I know where to look in the admin UI

I would expect us to use `description` for that. Seems analogous to how we log JobId a lot, and I've yet to ever actually use that to correlate anything, preferring Queue and/or Jobtype. :shrug: